CMakeLists.txt

```diff
@@ -1,17 +1,54 @@
-cmake_minimum_required(VERSION 2.8)
+cmake_minimum_required(VERSION 2.8.11)
 
 project(Zopfli)
 
-option(BUILD_SHARED_LIBS "Build Zopfli with shared libraries" OFF)
+# Check if being built standalone
+set(zopfli_standalone FALSE)
+get_directory_property(zopfli_parent_directory PARENT_DIRECTORY)
+if(zopfli_parent_directory STREQUAL "")
+  set(zopfli_standalone TRUE)
+endif()
+unset(zopfli_parent_directory)
+
+#
+# Options
+#
 
-if(NOT CMAKE_BUILD_TYPE)
-  set(CMAKE_BUILD_TYPE Release)
+# Default to BUILD_SHARED_LIBS if set, but allow overriding it
+set(zopfli_shared_default OFF)
+if(DEFINED BUILD_SHARED_LIBS)
+  set(zopfli_shared_default ${BUILD_SHARED_LIBS})
 endif()
+option(ZOPFLI_BUILD_SHARED "Build Zopfli with shared libraries" ${zopfli_shared_default})
```

Comment:
These lines look a bit as if `BUILD_SHARED_LIBS` and `ZOPFLI_BUILD_SHARED` could both be used to flip the same bit from the outside. Are both needed or can this be simplified to a single variable?

Reply:
They could indeed. The reason for the separate variable is again the potential use as a subproject.

BUILD_SHARED_LIBS is a global variable that causes all libraries that are not specifically static to be built shared.

If Zopfli is being built as a subproject, we should not change the state of this variable, potentially changing how other parts of the parent project are being built. On the other hand, unless told otherwise, we should respect the value of this variable if it has been set.

The logic here is an attempt to support this, and also to allow situations like building Zopfli as a static library as part of a build where BUILD_SHARED_LIBS is ON, or vice versa.

So, as you noted, this extra variable is redundant for building Zopfli on its own, but hopefully useful when built as a subproject.

Another option is to use a normal variable rather than an option for this, for instance BROTLI_BUNDLED_MODE in [Brotli](https://github.com/google/brotli/blob/master/CMakeLists.txt).

Reply:
Thanks for the explanation.